spanner/cloud-client/src/main/java/com/example/spanner/LoadCsvExample.java

```diff
@@ -0,0 +1,250 @@
+    ResultSet spannerType = connection.createStatement()
+        .executeQuery("SELECT spanner_type FROM information_schema.columns "
+            + "WHERE table_name = \"" + tableId + "\"");
+    while (spannerType.next()) {
+      spannerSchema.add(parseSpannerDataType(spannerType.getString("spanner_type")));
+    }
+  }
+
+  public static void parseTableHeaders(String tableId) throws SQLException {
+    ResultSet tableHeader = connection.createStatement()
+        .executeQuery("SELECT column_name FROM information_schema.columns "
+            + "WHERE table_name = \"" + tableId + "\"");
+    while (tableHeader.next()) {
+      headers.add(tableHeader.getString("column_name"));
+    }
+  }
+
+  public static boolean isValidHeaders(CSVParser parser) {
+    List<String> allHeaders = parser.getHeaderNames();
+    for (int i = 0; i < allHeaders.size(); i++) {
+      if (!allHeaders.get(i).equals(headers.get(i))) {
+        System.err.println(
+            "Header " + allHeaders.get(i) + " does not match database table header " + headers
+                .get(i));
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public static CSVFormat setFormat(CommandLine cmd) {
+    CSVFormat parseFormat;
+    if (cmd.hasOption("f")) {
+      switch (cmd.getOptionValue("f").toUpperCase()) {
+        case (EXCEL):
+          parseFormat = CSVFormat.EXCEL;
+          break;
+        case (POSTGRESQL_CSV):
+          parseFormat = CSVFormat.POSTGRESQL_CSV;
+          break;
+        case(POSTGRESQL_TEXT):
+          parseFormat = CSVFormat.POSTGRESQL_TEXT;
+          break;
+        default:
+          parseFormat = CSVFormat.DEFAULT;
+      }
+    } else {
+      parseFormat = CSVFormat.DEFAULT;
+    }
+
+    if (cmd.hasOption("n")) {
+      parseFormat = parseFormat.withNullString(cmd.getOptionValue("n"));
+    }
+
+    if (cmd.hasOption("d")) {
+      if (cmd.getOptionValue("d").length() != 1) {
+        throw new IllegalArgumentException("Invalid delimiter character entered.");
+      }
+      parseFormat = parseFormat.withDelimiter(cmd.getOptionValue("d").charAt(0));
+    }
+
+    if (cmd.hasOption("e")) {
+      if (cmd.getOptionValue("e").length() != 1) {
+        throw new IllegalArgumentException("Invalid escape character entered.");
+      }
+      parseFormat = parseFormat.withEscape(cmd.getOptionValue("e").charAt(0));
+    }
+
+    if (cmd.hasOption("h") &&  cmd.getOptionValue("h").equalsIgnoreCase("True")) {
+      parseFormat = parseFormat.withFirstRecordAsHeader();
+    }
+    return parseFormat;
+  }
+
+  public static void loadData(Iterable<CSVRecord> records, String tableId)
+      throws SQLException {
+    System.out.println("Writing data into table...");
+    List<Mutation> mutations = new ArrayList<>();
+    for (CSVRecord record : records) {
+      WriteBuilder builder = Mutation.newInsertOrUpdateBuilder(tableId);
+      for (int i = 0; i < headers.size(); i++) {
```

Comment:
One more feature that you could add, which will help your COPY work is:
- If there is no header specified, assume the CSV file contains all columns of the table.
- If there is a header, the CSV can contains a subset of columns (other columns will have the NULL value).

Could you also add a unit test with null strings? To make sure that these values are written as NULL in Spanner.

Reply:
Updated so that both those features are now added. Also added a test for parsing with the null strings.

Not sure if you want to unit tests to also involve directly making an executQuery call to Spanner to read back the values and directly compare then back again to the original values?